### PR TITLE
Implement New Identity Calculation

### PR DESCRIPTION
**Pull request by gallmann-ubique**

### CODE REVIEW

File: crowdnotifier-sdk/src/main/java/org/crowdnotifier/android/sdk/utils/CryptoUtils.java [hunk 1]
```diff
@@ -219,9 +223,8 @@ public NoncesAndNotificationKey getNoncesAndNotificationKey(ProtoV3.QRCodePayloa
...
 	public NoncesAndNotificationKey getNoncesAndNotificationKey(byte[] qrCodePayload) {
 		try {
-			byte[] hkdfOutput =
-					Hkdf.computeHkdf("HMACSHA256", qrCodePayload, new byte[0],
-							"CrowdNotifier_v3".getBytes(StandardCharsets.US_ASCII), 96);
+			byte[] hkdfOutput = Hkdf.computeHkdf("HMACSHA256", qrCodePayload, new byte[0],
+					"CrowdNotifier_v3".getBytes(StandardCharsets.US_ASCII), 96);
 			byte[] nonce1 = Arrays.copyOfRange(hkdfOutput, 0, 32);
 			byte[] nonce2 = Arrays.copyOfRange(hkdfOutput, 32, 64);
 			byte[] notificationKey = Arrays.copyOfRange(hkdfOutput, 64, 96);
```
Comment (ineiti) [inline crowdnotifier-sdk/src/main/java/org/crowdnotifier/android/sdk/utils/CryptoUtils.java:230]:
```suggestion
			byte[] noncePreId = Arrays.copyOfRange(hkdfOutput, 0, 32);
			byte[] nonceTimeKey = Arrays.copyOfRange(hkdfOutput, 32, 64);
			byte[] notifyKey = Arrays.copyOfRange(hkdfOutput, 64, 96);
```

**Review by ineiti — CHANGES_REQUESTED** [review]
I'd also add the following changes:
- `generateIdentityV3(byte[] qrCodePayload, long startOfInterval, int intervalLength)` - with a check that `intervalLength` is <= 86400
- rename `longToBytes` to `int64ToBytesBE` or `int64ToBytesBigEndian`
- rename `intToBytes` to `int32ToBytesBE` or `int32ToBytesBigEndian`

**Comment by gallmann-ubique** [comment]
> * `intervalLength` is <= 86400

Thanks for these good suggestions! I have implemented them.

File: crowdnotifier-sdk/src/main/java/org/crowdnotifier/android/sdk/utils/CryptoUtils.java [hunk 2]
```diff
@@ -195,22 +195,26 @@ public IBECiphertext encryptInternal(G2 masterPublicKey, byte[] identity, byte[]
...
 
-	public byte[] generateIdentityV3(ProtoV3.QRCodePayload qrCodePayload, long startOfInterval) {
-		return generateIdentityV3(qrCodePayload.toByteArray(), startOfInterval);
+	public byte[] generateIdentityV3(ProtoV3.QRCodePayload qrCodePayload, long startOfInterval, int intervalLength) {
+		return generateIdentityV3(qrCodePayload.toByteArray(), startOfInterval, intervalLength);
 	}
 
-	public byte[] generateIdentityV3(byte[] qrCodePayload, long startOfInterval) {
+	public byte[] generateIdentityV3(byte[] qrCodePayload, long startOfInterval, int intervalLength) {
+		if (intervalLength < 0 || intervalLength > 86400) throw new RuntimeException("intervalLength must be between 0 and 86400");
```
Comment (ineiti) [inline crowdnotifier-sdk/src/main/java/org/crowdnotifier/android/sdk/utils/CryptoUtils.java:207]: Discussing with @wouterl , we should have `intervalLength <= 900`.

**Review by wouterl — COMMENTED** [review]
FWIW, this implementation has the same somewhat icky overload of names (that happen to be the same) for the V2 version. Ok with me, assuming V2 will die soon.

**Pull request merged**